File: lib/features/feed/domain/models/comment.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'comment.freezed.dart';
part 'comment.g.dart';

@freezed
abstract class Comment with _$Comment {
  const factory Comment({
    @Default('') String id,
    @JsonKey(name: 'post_id') @Default('') String postId,
    @JsonKey(name: 'user_id') @Default('') String userId,
    @JsonKey(name: 'parent_comment_id') String? parentCommentId,
    @Default('user') String username,
    @JsonKey(name: 'user_avatar') @Default('') String userAvatar,
    @Default('') String content,
    @Default(0) @JsonKey(name: 'likes_count') int likes,
    @Default(0) @JsonKey(name: 'replies_count') int repliesCount,
    @Default(false) @JsonKey(name: 'is_liked') bool isLiked,
    @JsonKey(name: 'created_at') required DateTime timestamp,
  }) = _Comment;

  const Comment._();

  factory Comment.fromJson(Map<String, dynamic> json) =>
      _$CommentFromJson(_normalizeCommentJson(json));

  static Map<String, dynamic> _normalizeCommentJson(Map<String, dynamic> json) {
    final Map<String, dynamic> normalized = Map.from(json);

    // Handle nested profile data if present
    final profile = json['profiles'];
    if (profile != null && profile is Map<String, dynamic>) {
      normalized['username'] = profile['username'] ?? normalized['username'];
      normalized['user_avatar'] =
          profile['avatar_url'] ?? normalized['user_avatar'];
    }

    normalized['user_avatar'] =
        normalized['user_avatar'] ??
        json['user_avatar'] ??
        json['avatar_url'] ??
        '';
    normalized['likes_count'] = json['likes_count'] ?? json['likes'] ?? 0;
    normalized['replies_count'] = json['replies_count'] ?? json['replies'] ?? 0;
    normalized['is_liked'] = json['is_liked'] ?? false;
    normalized['created_at'] =
        json['created_at'] ?? DateTime.now().toIso8601String();
    return normalized;
  }
}

File: winui/src/Oasis.WinUI/Core/Models/Comment.cs
using System.Text.Json;

namespace Oasis.WinUI.Core.Models;

public sealed class Comment
{
    public string Id { get; set; } = "";
    public string PostId { get; set; } = "";
    public string UserId { get; set; } = "";
    public string UserName { get; set; } = "";
    public string UserAvatar { get; set; } = "";
    public string Content { get; set; } = "";
    public DateTime Timestamp { get; set; } = DateTime.Now;
    public long LikesCount { get; set; }
    public long RepliesCount { get; set; }
    public bool IsLiked { get; set; }
    public string? ParentCommentId { get; set; }
    public bool IsEdited { get; set; }

    public static Comment FromJson(JsonElement el)
    {
        return new Comment
        {
            Id = JsonUtil.S(el, "id") ?? "",
            PostId = JsonUtil.S(el, "post_id") ?? "",
            UserId = JsonUtil.S(el, "user_id") ?? "",
            UserName = JsonUtil.S(el, "username", "full_name")
                       ?? JsonUtil.NestedProfile(el, "profiles", "username") ?? "",
            UserAvatar = JsonUtil.S(el, "avatar_url", "user_avatar")
                         ?? JsonUtil.NestedProfile(el, "profiles", "avatar_url") ?? "",
            Content = JsonUtil.S(el, "content") ?? "",
            Timestamp = JsonUtil.Dt(el, "created_at", "timestamp") ?? DateTime.Now,
            LikesCount = JsonUtil.L(el, "likes_count", "likes"),
            RepliesCount = JsonUtil.L(el, "replies_count", "replies"),
            IsLiked = JsonUtil.B(el, "is_liked", "isLiked"),
            ParentCommentId = JsonUtil.S(el, "parent_comment_id"),
            IsEdited = JsonUtil.B(el, "is_edited"),
        };
    }

    public static List<Comment> FromJsonArray(string json)
    {
        var result = new List<Comment>();
        try
        {
            using var doc = JsonDocument.Parse(json);
            if (doc.RootElement.ValueKind != JsonValueKind.Array) return result;
            foreach (var el in doc.RootElement.EnumerateArray())
            {
                if (el.ValueKind == JsonValueKind.Object) result.Add(FromJson(el));
            }
        }
        catch
        {
            // ignore
        }
        return result;
    }
}
